src/webview/js/js.js

```diff
@@ -333,6 +372,23 @@ function someVisibleMessage(top: number, bottom: number): ?Element {
   );
 }
 
+/** Returns some message element which is both visible and read, if any. */
+function someVisibleReadMessage(top: number, bottom: number): ?Element {
+  function checkRead(candidate: ?Element): ?Element {
+    return candidate && isRead(candidate, top, bottom) ? candidate : null;
+  }
+
+  // Algorithm: If there's a visible message that isn't read, that means
+  // it's partway off the bottom of the screen. Therefore, either:
+  // * the message above it will be visible and read, or
+  // * there are no visible read messages.
+  const visible = someVisibleMessage(top, bottom);
+  if (visible === null || visible === undefined) {
+    return visible;
+  }
+  return checkRead(visible) || checkRead(previousMessage(visible));
+}
```

Comment:
Hmm, I know this is following the code I suggested, but rereading I see it isn't quite right! I think with a small fix it is, though.

Specifically, if we're in the middle of a long message, then we'll get that as `visible`; we'll try that in the first `checkRead`, and it'll fail; and then we'll find the previous message (supposing there is one) and pass that to `checkRead`. And it will accept it and we'll return that, even though it's out of view above the screen and isn't visible.

One fix that I believe does the job is to add an `isVisible` check in `checkRead` (and also change its name to match: like `checkVisibleAndRead`, or `checkQualified`, or just `check`.)

Alternatively, could inline the `checkRead` helper and just write out the sequence of conditionals, with something like `const previous = previousMessage(visible);` at the point where we're about to consider that one.

Reply:
Ah, yup. Updated to be `checkReadAndVisible`, since inlining it seemed like a pain.